src/main/java/org/jboss/threads/JBossThread.java

```diff
@@ -153,18 +160,32 @@ public void interrupt() {
         if (isInterrupted()) return;
         final AtomicInteger stateRef = this.stateRef;
         int oldVal, newVal;
-        do {
+        int spins = 0;
+        for (;;) {
             oldVal = stateRef.get();
-            if (oldVal == STATE_INTERRUPT_PENDING || oldVal == STATE_INTERRUPT_IN_PROGRESS) {
+            if (oldVal == STATE_INTERRUPT_PENDING) {
                 // already set
                 Messages.msg.tracef("Interrupting thread \"%s\" (already interrupted)", this);
                 return;
-            } else if (oldVal == STATE_INTERRUPT_DEFERRED) {
-                newVal = STATE_INTERRUPT_PENDING;
+            } else if (oldVal == STATE_INTERRUPT_IN_PROGRESS) {
+                // wait for interruption on other thread to be completed
+                if (spins < MAX_INTERRUPT_SPINS) {
```

Comment:
spin is not autoincremented on each loop, missed a spin++?

Reply:
You're right; fixed.